Move non-web related processing into processor, add "publish" package

roncohen opened the pull request:
This introduces the `processor/stream` and `publish` packages and moves relevant code from `beater` to the new packages.

roncohen commented:
jenkins, test this please

simitt commented:
jenkins, test this please

roncohen commented:
jenkins, test this please

roncohen commented on utility/request_time.go:
```diff
@@ -0,0 +1,42 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
```
I agree generally agree, however, `processor` needs to access the `requestTime` the context that is passed to it. Keeping this in `beater` creates a circular dependency. I could extract requestTime from the context before calling `HandleStream` and add it as yet another parameter to that method, but it seemed to belong well in the context

simitt replied:
As far as I could see `requestTime` was only used in the tests, where we could just add a time to context manually. But up to you if you want to move this.

roncohen replied:
Did you mean we should just use time.Now() in here instead?
https://github.com/elastic/apm-server/pull/1324/files#diff-9829d1c9b6183c7ba00ca10afb6376b9R422

I can also move it to the `stream` or `processing` package if you think that's better.

simitt replied:
Giving it a second thought, let's move it to `utility` for now and then generally investigate logic in this package when v2 is done.

roncohen commented on processor/stream/result.go:
```diff
@@ -131,7 +131,7 @@ func (s *streamResponse) String() string {
 }
 
-func (s *streamResponse) statusCode() int {
+func (s *Result) StatusCode() int {
```
good point. Do you agree that it makes sense to have `Result` be part of `stream` but translate it to a http type error in `beater` ?

simitt replied:
yes, that's exactly what I meant!

roncohen commented on beater/v2_handler_test.go:
```diff
@@ -18,313 +18,23 @@
-	}
-	expectedBuf, err := expected.marshal()
+func TestInvalidContentType(t *testing.T) {
+	req, err := http.NewRequest("POST", "/v2/intake", nil)
```
coming up next 👍

roncohen commented:
I moved the v2 integration tests into `stream` as you suggested and I'm using the approval system for the results now. I also simplified the error response and that made it possible to remove a lot of code 👍

At the moment, the stream result is just json encoded in `beater` and used as the http response. That means there's still some overlapping concerns, e.g. concerns are not completely separated. 

I didn't feel it would makes sense to create a new structure in `beater` because it would be a copy of `Result` in `stream` with some added json tags, so i put the tags directly on the `stream.Result`. Let me know what you think.

roncohen commented:
If it's OK with you, I’m planning to create the v2 server integration tests you asked for in a separate PR. This one is already doing too much.

simitt commented on processor/stream/stream_processor.go:
```diff
@@ -0,0 +1,239 @@
+						Type:    ShuttingDownErrType,
+						Message: "server is shutting down",
+					})
+				case publish.ErrFull:
```
As https://github.com/roncohen/apm-server/pull/3 has not been merged, I guess this will need to be implemented again after merging this PR.

roncohen replied:
I needed to refactor this area, so your PR wouldn't apply cleanly anymore. During the refactor i changed the behavior in this PR so that it returns when the queue is full, like in your PR.

roncohen commented on processor/stream/stream_processor.go:
```diff
@@ -0,0 +1,239 @@
+}
+
+// readBatch will read up to `batchSize` objects from the ndjson stream
+// it returns a slice of eventables, a serverResponse and a bool that indicates if we're at EOF.
```
ah yes, that comment didn't get updated :) 

Returning the response instead of taking a reference would requires us to merge the two responses every time readBatch returns. There's no merging code atm. Do you think it's worth it?

simitt replied:
Errors should be edge cases. Thus, I don't expect a big performance impact one way or the other. I personally find it cleaner, but no strong opinion.

roncohen replied:
I've updated the comment now. I prefer to leave the arguments for now, but happy to discuss again at a later point.

roncohen commented on processor/stream/stream_processor.go:
```diff
@@ -0,0 +1,239 @@
+				response.LimitedAdd(&Error{
+					Type:     InvalidInputErrType,
+					Message:  err.Error(),
+					Document: string(reader.LastLine()),
```
`rawModel` is a map[string]interface{} here, not a string or []byte.

roncohen commented on processor/stream/result.go:
```diff
@@ -0,0 +1,72 @@
+}
+
+func (r *Result) LimitedAdd(err error) {
+	if len(r.Errors) < errorsLimit {
```
to be clear, I'm happy to discuss this, but i think we should move forward with this behavior at this time

roncohen commented:
thanks for the thorough review @simitt !
